client/lib/use-local-storage/index.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { useCallback, useRef, useState } from 'react';
+
+export function useLocalStorage< T >(
+	storageKey: string,
+	initialValue: T | ( () => T )
+): [ T, ( newValue: T | ( ( oldValue: T ) => T ) ) => void ] {
+	const [ storedValue, setStoredValue ] = useState( () => {
+		const item = typeof window === 'undefined' ? null : window.localStorage.getItem( storageKey );
+		if ( item !== null ) {
+			try {
+				return JSON.parse( item ) as T;
```

Comment:
Should we make `serialize` and `deserialize` parameters (and default to the `JSON` methods?)

Reply:
Could be useful, but I'm not sure in what situations. A downside to coding a solution when I don't have an example of the problem it's solving yet.

Pro to allowing custom (de)serialize functions: it could be done more efficiently perhaps. If I know the state will always be a single string, it'd be more efficient to write it to local storage directly.

Con: it'd be easy to write a serialize function that ignores how to correctly encode `null` because you don't need that right now, and then a developer in the future starts using `null` in the state but doesn't realize the (de)serialize functions don't support it.
Would this be mitigated using Typescript and making sure the (de)serialize functions accept all values of type `T`.

It feels like something that could be added later after there was a specific use case.